components/action-mgt/org.wso2.carbon.identity.action.execution/src/main/java/org/wso2/carbon/identity/action/execution/ActionExecutorServiceImpl.java

```diff
@@ -0,0 +1,399 @@
+
+        return INSTANCE;
+    }
+
+    public ActionExecutionStatus execute(ActionType actionType, Map<String, Object> eventContext, String tenantDomain)
+            throws ActionExecutionException {
+
+        try {
+            List<Action> actions = getActionsByActionType(actionType, tenantDomain);
+            validateActions(actions, actionType);
+            ActionExecutionRequest actionRequest = buildActionExecutionRequest(actionType, eventContext);
+            ActionExecutionResponseProcessor actionExecutionResponseProcessor = getResponseProcessor(actionType);
+            return executeAction(actions.get(0), actionRequest, actionType, eventContext,
+                    actionExecutionResponseProcessor);
+        } catch (ActionExecutionRuntimeException e) {
+            // todo: add to diagnostics
+            LOG.error("Skip executing actions for action type: " + actionType.name() + ". Error: " + e.getMessage(), e);
+            return new ActionExecutionStatus(ActionExecutionStatus.Status.FAILURE, eventContext);
+
+        }
+    }
+
+    private List<Action> getActionsByActionType(ActionType actionType, String tenantDomain) throws
+            ActionExecutionRuntimeException {
+
+        try {
+            return ActionExecutionServiceComponentHolder.getInstance().getActionManagementService()
+                    .getActionsByActionType(Action.ActionTypes.valueOf(actionType.name()).getPathParam(), tenantDomain);
+        } catch (ActionMgtException e) {
+            throw new ActionExecutionRuntimeException("Error occurred while retrieving actions.", e);
+        }
+    }
+
+    private void validateActions(List<Action> actions, ActionType actionType) throws ActionExecutionException {
+
+        if (actions.isEmpty()) {
+            throw new ActionExecutionRuntimeException("No actions found for action type: " + actionType);
+        }
+        if (actions.size() > 1) {
+            // when multiple actions are supported for an action type the logic below needs to be improved such that,
+            // a successful processing from one action becomes the input to the successor.
+            throw new ActionExecutionException("Multiple actions found for action type: " + actionType.name() +
+                    ". Current implementation doesn't support multiple actions for a single action type.");
+        }
+    }
+
+    private ActionExecutionRequest buildActionExecutionRequest(ActionType actionType, Map<String, Object> eventContext)
+            throws ActionExecutionException {
+
+        ActionExecutionRequestBuilder requestBuilder =
+                ActionExecutionRequestBuilderFactory.getActionExecutionRequestBuilder(actionType);
+        if (requestBuilder == null) {
+            throw new ActionExecutionException("No request builder found for action type: " + actionType);
+        }
+        try {
+            return requestBuilder.buildActionExecutionRequest(actionType, eventContext);
```

Comment:
Why do we need to specify the actionType here given that requestBuilder has getSupportedActionType() which returns actionType ?

Reply:
replied on earlier comment. Will fix it